### PR TITLE
Attempt at getting a non-centered parameterization working

### PR DESCRIPTION
**Pull request by bbbales2**

### CODE REVIEW

File: model.py [hunk 1]
```diff
@@ -62,19 +62,23 @@ def lpdf(unconstrained_parameter_vector):
...
+                    parameter, constraints_jacobian_adjustment = constraints.lower(variable, lower)
                 elif lower == float("inf") and upper < float("inf"):
-                    parameter, jacobian_adjustment = constraints.upper(variable, upper)
+                    parameter, constraints_jacobian_adjustment = constraints.upper(variable, upper)
                 elif lower > float("inf") and upper < float("inf"):
-                    parameter, jacobian_adjustment = constraints.finite(variable, lower, upper)
+                    parameter, constraints_jacobian_adjustment = constraints.finite(variable, lower, upper)
                 
+                if centered == False:
+                    parameter, noncentering_jacobian_adjustment = constraints.offset_multiply(variable, mu, sd)
```
[the remark this answers is not in the document]
Comment (Dashadower) [reply model.py:76]: So it seems like we need to keep the values of the evaluated inner expressions, right?

[the remark this answers is not in the document]
Comment (Dashadower) [reply model.py:76]: P.S. the centering syntax is concise and simple which is great, but how about we allow users to define non-centered parameterizations themselves through statements, just like stan?

[the remark this answers is not in the document]
Comment (bbbales2) [reply model.py:76]: > a variable map which will keep the variable values

It's annoying that the variables of concern are kind of unnamed temporaries with the centered=True syntax.

> how about we allow users to define non-centered parameterizations themselves through statements

Yeah, this makes sense.

If we have an equals operator then maybe we could do:
```
mu[group] = ...;
sd[group] = ...;
param<offset = mu, multiplier = sd>[group] ~ normal(mu, sd);
```

The reason I named the indices on the left hand side of =s is that if we want to use those indices on the right hand side (we might) they need unambiguous names.

And so here mu and sd are calculated once and saved in a global or whatever and used twice.

Here's another attempt that could be interesting:
```
y ~ normal(param[group], sigma);
param[group] = scale * param_z[group] + offset;
param_z[group] ~ normal(0, 1)
scale = ...;
offset = ...;
```

Here the first line doesn't find param, and so deduces it must exist with a certain index.

The second line will figure out we make param by assignment -- and then scale, parma_z, and offset aren't defined yet and so the compiler can deduce they need to exist somewhere with certain shapes (scale and offset are both scalar, cause no index, and param_z's index matches param's).

Then the param_z thing works as usual.

[the remark this answers is not in the document]
Comment (Dashadower) [reply model.py:76]: Yeah the second code snippet is what I was thinking of. 

>  the compiler can deduce they need to exist somewhere with certain shapes 
This looks like it's completely possible with 2 passes of the code:
1. resolve ambiguities
2. verify shapes match for statements

[the remark this answers is not in the document]
Comment (bbbales2) [reply model.py:76]: Yeah multiple passes seems good.

[the remark this answers is not in the document]
Comment (bbbales2) [reply model.py:76]: Seems like with that setup it'd also be possible to automatically transform something in the centered form:

```
param[group] ~ normal(mu, sd);
```

to the non-centered form.

**Comment by Dashadower** [comment]
(still work in progress, will mention when it's ready)

This branch pretty much tries to make assignments work, which enables hierarchical non-centered parameterization in the form of:

y[n] ~ normal(theta[n], sigma);
theta[n] = mu + tau * unit_normal[n];

unit_normal[n] = normal(0, 1);
tau<lower=0> ~ normal(0, 2);
mu ~ normal(0, 5);
sigma = 1;

**Comment by Dashadower** [comment]
Oh and is it fine if we change the target branch to main from lag?

**Comment by Dashadower** [comment]
Status update:

I'm not exactly sure how well it will handle for more complex conditions. I tried testing with the following modified skills model:

```
score_diff~normal(skills[home_team, year]-skills[away_team, year],sigma_mod[year]);
skills[team, year] ~ normal(skills_mu[year], 1);
skills_mu[year] ~ normal(0.0, 1.0);
sigma_mod[year] = sigma[year];
sigma<lower=0.0>[year] ~ normal(0.0, 10.0);
```

And it seems like it works?

However there are a couple of problems.

1. Can't retrieve assigned variables. After we get the draws dataframe, we would have to manually transform by executing the code, which leads the problem 2
2. Since our `code()` function works on indexed values, it's extremely hard to extract the original expression that doesn't have subscripts tacked on. 
3. As of now, when evaluating assignments, it correctly works if and only if all the subscripts on that statement line to be the same. So for the above model, the numpy shape of `sigma_mod` and `sigma` should be the same. The reason for this is because when we evaluate the statement, it strips the brackets `[year]` and just accesses the numpy array, does operations and assigns it as lhs(`sigma_mod`) with `eval`. So you can see there are lots of ways this can go wrong.

**Comment by bbbales2** [comment]
> when evaluating assignments, it correctly works if and only if all the subscripts on that statement line are the same

Oh interesting. I think this should generalize but I'll dig in.

> Since our code() function works on indexed values, it's extremely hard to extract the original expression that doesn't have subscripts tacked on.

Yeah I think I do a trick for that lol.

File: rat/variables.py [hunk 2]
```diff
@@ -148,6 +164,7 @@ class IndexUse:
     shifts: Tuple[Union[str, None]] = None
 
     def to_numpy(self):
+        self.df = self.df.loc[:, self.names]
```
Comment (bbbales2) [inline rat/variables.py:167]: Any memory of where this came from?

Comment (Dashadower) [reply rat/variables.py:167]: So `variable.Index.base_df` is the "reference dataframe", which contains the unique subscripts sets and assigns an index to each of them. In `variable.IndexUse`, we receive some other subscript dataframe, which contains subscripts that are being used in that row. `to_numpy()` would get the reference indexes of the subscript dataframe.

That line in particular, say we have `skills[team, year] ~ normal(skills_mu[year], tau);`. 
It takes the index dataframe of `skills_mu`(`variable.IndexUse.df` of `skills_mu`), which would be defined by the reference dataframe of `skills`(`Index.base_df` of `skills`) and extract only the column that `skills_mu` use, which would be `year`.

Once that's done, it compares that to the reference dataframe of itself, yielding numpy indexes of the unique subscripts that are within the index dataframe.

File: rat/variables.py [hunk 3]
```diff
@@ -95,6 +106,27 @@ def get_numpy_indices(self, df):
             "__index"
         ].to_numpy(dtype=int)
 
+    def log_summary(self, log_level=logging.INFO):
+        for index_num in range(len(self.indexed_sets)):
+            logging.log(log_level, f"Subscript {index_num} - defined as union of the following columns({','.join(self.indexed_sets[index_num])}), with column name '{self.df.columns[index_num]}'")
+
+    def check_and_return_index(self, index_key: Tuple[str]) -> Tuple[str]:
```
Comment (Dashadower) [inline rat/variables.py:116]: @bbbales2 
This commit added this function, where I intended to make `variables.Index` keep track of the subscripts and dataframe column names. For example for the baseball model, we subscript `skills` in 3 different ways (`skills[home_team, year]`, `skills[away_team, year]`, `skills[team, year]`), which kinda confuses what names we should set the underlying df(`self.df`)to be and when loc-ing them. 
Previously on the original implementation, we just used the column names of the first indexing call, but actually keeping track the different combinations of how the subscripts called gives us better opportunity for maintenance and improvement.

On an added plus, it allows for some cool info output:
```
INFO:root:rat compiler - starting code generation for 6 line(s) of code.
INFO:root:Subscript data for parameter skills:
INFO:root:Subscript 0 - defined as union of the following columns(home_team,away_team), with column name 'team'
INFO:root:Subscript 1 - defined as union of the following columns(year), with column name 'year'
INFO:root:-----
INFO:root:Subscript data for parameter sigma_mod_plus:
INFO:root:Subscript 0 - defined as union of the following columns(year), with column name 'year'
INFO:root:-----
INFO:root:Subscript data for parameter skills_mu:
INFO:root:Subscript 0 - defined as union of the following columns(year), with column name 'year'
INFO:root:-----
INFO:root:Subscript data for parameter sigma_mod:
INFO:root:Subscript 0 - defined as union of the following columns(year), with column name 'year'
INFO:root:-----
INFO:root:Subscript data for parameter sigma:
INFO:root:Subscript 0 - defined as union of the following columns(year), with column name 'year'
INFO:root:-----
```

**Review by Dashadower — APPROVED** [review]
This is probably not perfect but it's a big feature rolling in!

File: rat/variables.py (hunk 3 above)
Comment (bbbales2) [reply rat/variables.py:116]: Looks good!

**Pull request merged**